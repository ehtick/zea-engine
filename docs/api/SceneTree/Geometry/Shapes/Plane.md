<a name="Plane"></a>

### Plane 
A class for generating a plane geometry.


**Extends**: <code>Mesh</code>  
<a name="new_Plane_new"></a>

### new Plane
Create a plane.


| Param | Type | Default | Description |
| --- | --- | --- | --- |
| SizeX | <code>number</code> | <code>1</code> | The length of the plane along the X axis. |
| SizeY | <code>number</code> | <code>1</code> | The length of the plane along the Y axis. |
| DetailX | <code>number</code> | <code>1</code> | The number of divisions along the X axis. |
| DetailY | <code>number</code> | <code>1</code> | The number of divisions along the Y axis. |
| addNormals | <code>boolean</code> | <code>true</code> | The addNormals value. |
| addTextureCoords | <code>boolean</code> | <code>true</code> | The addTextureCoords value. |

