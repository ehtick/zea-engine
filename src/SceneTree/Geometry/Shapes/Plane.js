import { Vec2 } from '../../../Math/Vec2'
import { Vec3 } from '../../../Math/Vec3'
import { NumberParameter } from '../../Parameters/index'
import { Registry } from '../../../Registry'
import { ProceduralMesh } from './ProceduralMesh'

/**
 * A class for generating a plane geometry.
 *
 * ```
 * const plane = new Plane(2.0, 1.5, 10, 10)
 * ```
 *
 * **Parameters**
 * * **SizeX(`NumberParameter`):** Length of the plane along `X` axis.
 * * **SizeY(`NumberParameter`):** Length of the plane along `Y` axis.
 * * **DetailX(`NumberParameter`):** Number of divisions along `X`axis.
 * * **DetailY(`NumberParameter`):** Number of divisions along `Y`axis.
 *
 * @extends {ProceduralMesh}
 */
class Plane extends ProceduralMesh {
  /**
   * Create a plane.
   * @param {number} [SizeX=1.0] - The length of the plane along the X axis.
   * @param {number} [SizeY=1.0] - The length of the plane along the Y axis.
   * @param {number} [DetailX=1] - The number of divisions along the X axis.
   * @param {number} [DetailY=1] - The number of divisions along the Y axis.
   * @param {boolean} [addNormals=true] - The addNormals value.
   * @param {boolean} [addTextureCoords=true] - The addTextureCoords value.
   */
  constructor(SizeX = 1.0, SizeY = 1.0, DetailX = 1, DetailY = 1, addNormals = true, addTextureCoords = true) {
    super()

    if (isNaN(SizeX) || isNaN(SizeY) || isNaN(DetailX) || isNaN(DetailY)) throw new Error('Invalid geom args')

    this.__sizeXParam = this.addParameter(new NumberParameter('SizeX', SizeX))
    this.__sizeYParam = this.addParameter(new NumberParameter('SizeY', SizeY))
    this.__detailXParam = this.addParameter(new NumberParameter('DetailX', DetailX))
    this.__detailYParam = this.addParameter(new NumberParameter('DetailY', DetailY))
    if (addNormals) this.addVertexAttribute('normals', Vec3)
    if (addTextureCoords) this.addVertexAttribute('texCoords', Vec2)
    this.__rebuild()

    const resize = () => {
      this.__resize()
    }
    const rebuild = () => {
      this.__rebuild()
    }
    this.__sizeXParam.on('valueChanged', resize)
    this.__sizeYParam.on('valueChanged', resize)
    this.__detailXParam.on('valueChanged', rebuild)
    this.__detailYParam.on('valueChanged', rebuild)
  }

  /**
   * The __rebuild method.
   * @private
   */
  __rebuild() {
    const detailX = this.__detailXParam.getValue()
    const detailY = this.__detailYParam.getValue()
    this.setNumVertices((detailX + 1) * (detailY + 1))
    this.setFaceCounts([0, detailX * detailY])

    let quadId = 0
    for (let i = 0; i < detailY; i++) {
      for (let j = 0; j < detailX; j++) {
        const v0 = (detailX + 1) * (i + 1) + j
        const v1 = (detailX + 1) * (i + 1) + (j + 1)
        const v2 = (detailX + 1) * i + (j + 1)
        const v3 = (detailX + 1) * i + j
        this.setFaceVertexIndices(quadId, [v0, v1, v2, v3])
        quadId = quadId + 1
      }
    }

    let voff = 0
    const normals = this.getVertexAttribute('normals')
    if (normals) {
      for (let i = 0; i <= detailY; i++) {
        for (let j = 0; j <= detailX; j++) {
          normals.getValueRef(voff).set(0, 0, 1)
          voff++
        }
      }
    }

    voff = 0
    const texCoords = this.getVertexAttribute('texCoords')
    if (texCoords) {
      for (let i = 0; i <= detailY; i++) {
        const y = i / detailY
        for (let j = 0; j <= detailX; j++) {
          const x = j / detailX
          texCoords.getValueRef(voff).set(x, y)
          voff++
        }
      }
    }

    this.__resize(false)
    this.emit('geomDataTopologyChanged', {})
  }

  /**
   * The __resize method.
   *
   * @private
   * @param {boolean} [emit=true] - If `true` emits `geomDataChanged` event.
   */
  __resize(emit = true) {
    const sizeX = this.__sizeXParam.getValue()
    const sizeY = this.__sizeYParam.getValue()
    const detailX = this.__detailXParam.getValue()
    const detailY = this.__detailYParam.getValue()
    const positions = this.getVertexAttribute('positions')
    let voff = 0
    for (let i = 0; i <= detailY; i++) {
      const y = (i / detailY - 0.5) * sizeY
      for (let j = 0; j <= detailX; j++) {
        const x = (j / detailX - 0.5) * sizeX
        positions.getValueRef(voff).set(x, y, 0.0)
        voff++
      }
    }

    this.setBoundingBoxDirty()
    if (emit) this.emit('geomDataChanged', {})
  }
}

Registry.register('Plane', Plane)
export { Plane }
