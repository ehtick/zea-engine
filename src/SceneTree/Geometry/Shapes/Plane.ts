import { NumberParameter } from '../../Parameters/index'
import { Registry } from '../../../Registry'
import { ProceduralMesh } from './ProceduralMesh'
import { Vec3Attribute } from '../Vec3Attribute'
import { Vec2Attribute } from '../Vec2Attribute'

/**
 * A class for generating a plane geometry.
 *
 * ```
 * const plane = new Plane(2.0, 1.5, 10, 10)
 * ```
 *
 * **Parameters**
 * * **SizeX(`NumberParameter`):** Length of the plane along `X` axis.
 * * **SizeY(`NumberParameter`):** Length of the plane along `Y` axis.
 * * **DetailX(`NumberParameter`):** Number of divisions along `X`axis.
 * * **DetailY(`NumberParameter`):** Number of divisions along `Y`axis.
 *
 * @extends {ProceduralMesh}
 */
class Plane extends ProceduralMesh {
  protected __detailXParam: NumberParameter
  protected __detailYParam: NumberParameter
  protected __sizeXParam: NumberParameter
  protected __sizeYParam: NumberParameter
  // topologyParams: string[]

  /**
   * Create a plane.
   * @param {number} [SizeX=1.0] - The length of the plane along the X axis.
   * @param {number} [SizeY=1.0] - The length of the plane along the Y axis.
   * @param {number} [DetailX=1] - The number of divisions along the X axis.
   * @param {number} [DetailY=1] - The number of divisions along the Y axis.
   * @param {boolean} [addNormals=true] - The addNormals value.
   * @param {boolean} [addTextureCoords=true] - The addTextureCoords value.
   */
  constructor(SizeX = 1.0, SizeY = 1.0, DetailX = 1, DetailY = 1, addNormals = true, addTextureCoords = true) {
    super()
    this.topologyParams = []
    if (isNaN(SizeX) || isNaN(SizeY) || isNaN(DetailX) || isNaN(DetailY)) throw new Error('Invalid geom args')

    this.__sizeXParam = this.addParameter(new NumberParameter('SizeX', SizeX)) as NumberParameter
    this.__sizeYParam = this.addParameter(new NumberParameter('SizeY', SizeY)) as NumberParameter
    this.__detailXParam = this.addParameter(new NumberParameter('DetailX', DetailX)) as NumberParameter
    this.__detailYParam = this.addParameter(new NumberParameter('DetailY', DetailY)) as NumberParameter
    if (addNormals) this.addVertexAttribute('normals', new Vec3Attribute())
    if (addTextureCoords) this.addVertexAttribute('texCoords', new Vec2Attribute())

    this.topologyParams.push('DetailX')
    this.topologyParams.push('DetailY')
  }

  /**
   * The rebuild method.
   * @private
   */
  rebuild(): void {
    const detailX = this.__detailXParam.getValue() || 1
    const detailY = this.__detailYParam.getValue() || 1
    this.setNumVertices((detailX + 1) * (detailY + 1))
    this.setFaceCounts([0, detailX * detailY])

    let quadId = 0
    for (let i = 0; i < detailY; i++) {
      for (let j = 0; j < detailX; j++) {
        const v0 = (detailX + 1) * (i + 1) + j
        const v1 = (detailX + 1) * (i + 1) + (j + 1)
        const v2 = (detailX + 1) * i + (j + 1)
        const v3 = (detailX + 1) * i + j
        this.setFaceVertexIndices(quadId, [v0, v1, v2, v3])
        quadId = quadId + 1
      }
    }

    let voff = 0
    const normals = <Vec3Attribute>this.getVertexAttribute('normals')
    if (normals) {
      for (let i = 0; i <= detailY; i++) {
        for (let j = 0; j <= detailX; j++) {
          normals.getValueRef(voff).set(0, 0, 1)
          voff++
        }
      }
    }

    voff = 0
    const texCoords = <Vec2Attribute>this.getVertexAttribute('texCoords')
    if (texCoords) {
      for (let i = 0; i <= detailY; i++) {
        const y = i / detailY
        for (let j = 0; j <= detailX; j++) {
          const x = j / detailX
          texCoords.getValueRef(voff).set(x, y)
          voff++
        }
      }
    }

    this.resize()
  }

  /**
   * The resize method.
   *
   * @private
   */
  resize(): void {
    const sizeX = this.__sizeXParam.getValue() || 1.0
    const sizeY = this.__sizeYParam.getValue() || 1.0
    const detailX = this.__detailXParam.getValue() || 1
    const detailY = this.__detailYParam.getValue() || 1
    const positions = <Vec3Attribute>this.getVertexAttribute('positions')
    if (!positions) return

    let voff = 0
    for (let i = 0; i <= detailY; i++) {
      const y = (i / detailY - 0.5) * sizeY
      for (let j = 0; j <= detailX; j++) {
        const x = (j / detailX - 0.5) * sizeX
        positions.getValueRef(voff).set(x, y, 0.0)
        voff++
      }
    }
  }
}

Registry.register('Plane', Plane)
export { Plane }
