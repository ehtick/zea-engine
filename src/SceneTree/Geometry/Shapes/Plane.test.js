import { Plane } from './Plane'

describe('Plane', () => {
  it('tests default parameters', () => {
    const plane = new Plane()

    expect(plane.getParameter('SizeX').getValue()).toBe(1)
    expect(plane.getParameter('SizeY').getValue()).toBe(1)
    expect(plane.getParameter('DetailX').getValue()).toBe(1)
    expect(plane.getParameter('DetailY').getValue()).toBe(1)
  })

  it('updates parameters', () => {
    const plane = new Plane()
    plane.getParameter('SizeX').setValue(3)
    plane.getParameter('SizeY').setValue(3)
    plane.getParameter('DetailX').setValue(3)
    plane.getParameter('DetailY').setValue(4)

    expect(plane.getParameter('SizeX').getValue()).toBe(3)
    expect(plane.getParameter('SizeY').getValue()).toBe(3)
    expect(plane.getParameter('DetailX').getValue()).toBe(3)
    expect(plane.getParameter('DetailY').getValue()).toBe(4)
  })

  // There's an issue with flags, that are preventing parameters to be exported.
  it('saves to JSON (serialization).', () => {
    const plane = new Plane(2, 2, 2, 2, true, true)
    const outputJSON = plane.toJSON()

    expect(outputJSON).toMatchSnapshot()
  })

  it('restores from JSON (serialization).', () => {
    const plane = new Plane()
    const inputJSON = {
      params: { SizeX: { value: 2 }, SizeY: { value: 2 }, DetailX: { value: 2 }, DetailY: { value: 2 } },
      type: 'Plane',
      numVertices: 9,
      vertexAttributes: {
        positions: {
          data: [-1, -1, 0, 0, -1, 0, 1, -1, 0, -1, 0, 0, 0, 0, 0, 1, 0, 0, -1, 1, 0, 0, 1, 0, 1, 1, 0],
          dataType: 'Vec3',
          defaultValue: 0,
          length: 9,
          splits: {},
          splitValues: [],
        },
        normals: {
          data: [0, 0, 1, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0, 0, 1],
          dataType: 'Vec3',
          defaultValue: 1.7976931348623157e308,
          length: 9,
          splits: {},
          splitValues: [],
        },
        texCoords: {
          data: [0, 0, 0.5, 0, 1, 0, 0, 0.5, 0.5, 0.5, 1, 0.5, 0, 1, 0.5, 1, 1, 1],
          dataType: 'Vec2',
          defaultValue: 1.7976931348623157e308,
          length: 9,
          splits: {},
          splitValues: [],
        },
      },
      faceCounts: [0, 4],
      faceVertexIndices: [3, 4, 1, 0, 4, 5, 2, 1, 6, 7, 4, 3, 7, 8, 5, 4],
    }
    plane.fromJSON(inputJSON)

    const newPlane = new Plane(2, 2, 2, 2, true, true)
    expect(plane.toJSON()).toEqual(newPlane.toJSON())
  })
})
