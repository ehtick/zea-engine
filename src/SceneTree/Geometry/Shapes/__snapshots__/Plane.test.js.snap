// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Plane saves to JSON (serialization). 1`] = `
Object {
  "params": Object {
    "DetailX": Object {
      "value": 2,
    },
    "DetailY": Object {
      "value": 2,
    },
    "SizeX": Object {
      "value": 2,
    },
    "SizeY": Object {
      "value": 2,
    },
  },
  "type": "Plane",
  "vertexAttributes": Object {},
}
`;
